src/workerd/api/queue.c++

```diff
@@ -133,4 +133,264 @@ kj::Promise<void> WorkerQueue::sendBatch(
   }).attach(kj::mv(client));
 };
 
+QueueMessage::QueueMessage(
+    v8::Isolate* isolate, rpc::QueueMessage::Reader message, IoPtr<QueueEventResult> result)
+    : id(kj::str(message.getId())),
+      timestamp(message.getTimestampNs() * kj::NANOSECONDS + kj::UNIX_EPOCH),
+      body(isolate, jsg::Deserializer(isolate, message.getData()).readValue()),
+      result(result) {}
+// Note that we must make deep copies of all data here since the incoming Reader may be
+// deallocated while JS's GC wrappers still exist.
+
+QueueMessage::QueueMessage(
+    v8::Isolate* isolate, IncomingQueueMessage message, IoPtr<QueueEventResult> result)
+    : id(kj::mv(message.id)),
+      timestamp(message.timestamp),
+      body(isolate, jsg::Deserializer(isolate, kj::mv(message.body)).readValue()),
+      result(result) {}
+
+jsg::Value QueueMessage::getBody(v8::Isolate* isolate) {
+  return body.addRef(isolate);;
+}
+
+void QueueMessage::retry() {
+  if (result->retryAll) {
+    return;
+  }
+
+  if (result->ackAll) {
+    auto msg = kj::str(
+        "Received a call to retry() on message ", id, " after ackAll() was already called. "
+        "Calling retry() on a message after calling ackAll() has no effect.");
+    IoContext::current().logWarning(msg);
+    return;
+  }
+
+  if (result->explicitAcks.contains(id)) {
+    auto msg = kj::str(
+        "Received a call to retry() on message ", id, " after ack() was already called. "
+        "Calling retry() on a message after calling ack() has no effect.");
+    IoContext::current().logWarning(msg);
```

Comment:
Nit: might be worthwhile to move the warning into a utility method that can be reused to reduce duplication in these methods.

Reply:
Possibly, although I'm just gonna leave it for now to keep changes to a minimum.